dawn/src/dawn/CodeGen/Cuda-ico/CudaIcoCodeGen.cpp

```diff
@@ -165,6 +176,38 @@ void CudaIcoCodeGen::generateRunFun(
       // fields used in the stencil
       const auto fields = support::orderMap(stage->getFields());
 
+      // lets figure out how many k levels we need to consider
+      std::stringstream k_size;
+      DAWN_ASSERT_MSG(intervalsConsistent(*stage),
+                      "intervals in a stage must have same bounds for now!\n");
+      auto interval = stage->getChild(0)->getInterval();
+      if(interval.levelIsEnd(iir::Interval::Bound::upper)) {
+        k_size << "kSize_ + " << interval.upperOffset() << " - "
+               << (interval.lowerOffset() + interval.lowerLevel());
+      } else {
+        k_size << interval.upperLevel() << " + " << interval.upperOffset() << " - "
+               << (interval.lowerOffset() + interval.lowerLevel());
+      }
+
+      // lets build correct block size
+      // runFun.addStatement("int dK = (kSize_ + LEVELS_PER_THREAD - 1) / LEVELS_PER_THREAD");
+      for(auto stageLoc : stageLocType) {
+        switch(stageLoc) {
+        case ast::LocationType::Cells:
+          runFun.addStatement("dim3 dG" + std::to_string(stage->getStageID()) + " = grid(" +
+                              k_size.str() + ", mesh_.NumCells)");
+          break;
+        case ast::LocationType::Edges:
+          runFun.addStatement("dim3 dG" + std::to_string(stage->getStageID()) + " = grid(" +
+                              k_size.str() + ", mesh_.NumEdges)");
+          break;
+        case ast::LocationType::Vertices:
+          runFun.addStatement("dim3 dG" + std::to_string(stage->getStageID()) + " = grid(" +
+                              k_size.str() + ", mesh_.NumVertices)");
+          break;
```

Comment:
Since only `mesh_.Num*` changes you could consider moving the call to `addStatement` outside the switch and computing just the string `mesh_.Num*` inside.